feat(SwapModal): Integrate new Sign and Approve modals

caybro opened the pull request:
### What does the PR do

- remove old popup and related components/adaptors
- simplify the API of the Sign/Approve modals to take pre-formatted amounts as strings
- correct login/auth icon
- SB: make it possible to open the new secondary modals
- SB: fix the `fetchSuggestedRoutes` response handling, pass around the `uuid` and inject it to the mocked suggested routes
- adjust SB pages and tests

Fixes #15443

### Affected areas

SwapModal

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

Approve:
![image](https://github.com/status-im/status-desktop/assets/5377645/6d58dfc0-1bfa-4e1d-a426-d15249f44289)

Sign:
![image](https://github.com/status-im/status-desktop/assets/5377645/7a14d172-30da-4eb2-a128-03429b7ecd6b)

caybro commented on storybook/pages/SwapModalPage.qml:
```diff
@@ -45,6 +46,12 @@ SplitView {
         }
     }
 
+    Popups {
```
@Khushboo-dev-cpp this is needed for the secondary popups to work here in SB

caybro commented on storybook/pages/SwapModalPage.qml:
```diff
@@ -78,6 +85,7 @@ SplitView {
                               accountTo, "amount = ", amount, " tokenFrom = ", tokenFrom, " tokenTo = ", tokenTo,
                               " disabledFromChainIDs = ", disabledFromChainIDs, " disabledToChainIDs = ", disabledToChainIDs,
                               " preferredChainIDs = ", preferredChainIDs, " sendType =", sendType, " lockedInAmounts = ", lockedInAmounts)
+                d.uuid = uuid
```
This was missing for tracking the suggested routes

caybro commented on storybook/pages/SwapModalPage.qml:
```diff
@@ -347,23 +353,29 @@ SplitView {
                 onClicked: {
-                    dSwapStore.suggestedRoutesReady(d.dummySwapTransactionRoutes.txNoRoutes)
+                    const txRoutes = d.dummySwapTransactionRoutes.txNoRoutes
+                    txRoutes.uuid = d.uuid
```
Here (and below) we need to inject the same `uuid`

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | e219fe9d | [#1](https://ci.status.im/job/status-desktop/job/prs/job/macos/job/aarch64/job/package/job/PR-15530/1/) | 2024-07-10 18:50:34 | ~6 min | `macos/aarch64` | [:apple:`dmg`](https://status-im-desktop-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-240710-184405-e219fe-pr15530-aarch64.dmg) |

Khushboo-dev-cpp commented on ui/app/AppLayouts/Wallet/panels/SwapInputPanel.qml:
```diff
@@ -58,6 +58,7 @@ Control {
     readonly property int rawValueMultiplierIndex: amountToSendInput.multiplierIndex
     readonly property bool valueValid: amountToSendInput.inputNumberValid
     readonly property bool amountEnteredGreaterThanBalance: value > maxSendButton.maxSafeValue
+    readonly property string accountBalanceFormatted: root.currencyStore.formatCurrencyAmount(d.maxCryptoBalance, d.selectedHolding.symbol)
```
this being exposed from here is rather weird, this is an account property should not really come from the SwapInputPanel, but from the adaptor instead.

caybro replied:
Yeah, was just discussing that with @alexjba ; looks like we have an issue there with the `ModelEntry` (pointing to the currently selected account) which returns old/incorrect values

caybro replied:
TLDR, I agree :)

I'll do some experiment/test case that exhibits that faulty behavior so that Alex can have a look at it; meanwhile I'd keep this (with a TODO/FIXME) as the `ModelEntry` approach was just returning wrong results for me (I had this code originally there)

Khushboo-dev-cpp replied:
ok the please please create a GH issue and link it here :)

caybro replied:
@alexjba see the video here:

[Záznam obrazovky z 2024-07-12 00-56-19.webm](https://github.com/user-attachments/assets/528f3cf8-2ed8-4311-91c8-da85221b8889)

The code is in SwapModalAdaptor, I even added the helper for chainId:
```qml
...
FastExpressionRole {
  name: "accountBalance"
  expression: {
    root.swapFormData.selectedNetworkChainId
    return d.processAccountBalance(model.address)
  }
  expectedRoles: ["address"]
  }
...
ModelEntry {
    id: selectedAccountEntry
    sourceModel: root.nonWatchAccounts
    key: "address"
    value: root.swapFormData.selectedAccountAddress
}
```

and then in SwapModal:
```qml
SwapApproveCapModal {
...
accountBalanceFormatted: root.swapAdaptor.selectedAccount.accountBalance.formattedBalance
```

Simply shows the "old" value from the previously selected network

caybro replied:
Not sure it's the `ModelEntry` or because we calculate it using a function which uses the old/deprecated `SubmodelProxyModel`

caybro replied:
I created a separate issue for this under: https://github.com/status-im/status-desktop/issues/15554

Khushboo-dev-cpp replied:
thanks @caybro can you please link the task in the code?

Khushboo-dev-cpp commented on ui/StatusQ/src/StatusQ/Core/Utils/AmountsArithmetic.qml:
```diff
@@ -86,12 +86,12 @@ QtObject {
      */
     function fromString(numStr) {
-        console.assert(typeof numStr === "string")
+        console.assert(typeof numStr === "string" || typeof numStr === "number")
```
this function clearly says fromString, I dont think we should do this here. @micieslak thoughts?

caybro replied:
That CTOR accepts more than a string, there's no reason to `assert` here :) See https://mikemcl.github.io/big.js/#big

caybro replied:
The naming was wrong from the start, we just didn't notice

Khushboo-dev-cpp replied:
but its the first time we dont pass a string but a number to this function so we are extending the function?

caybro replied:
No, not really extending but correcting the expectations I'd say... not sure, I can remove it if you insist :)

Khushboo-dev-cpp commented on ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:
```diff
@@ -271,18 +272,15 @@ QObject {
-
-        let account = selectedAccountEntry.item
-        let accountAddress = account.address
+        const accountAddress = root.selectedAccount.address
```
we should add null/undefined check when using selectedAccount

caybro replied:
OK, can add

caybro replied:
Just used the account address directly, we don't have to take the detour :facepalm:

Khushboo-dev-cpp commented on ui/app/AppLayouts/Wallet/popups/swap/SwapModalAdaptor.qml:
```diff
@@ -271,18 +272,15 @@ QObject {
     function sendSwapTx() {
-        let account = selectedAccountEntry.item
-        let accountAddress = account.address
+        const accountAddress = root.selectedAccount.address
```
same as above

Khushboo-dev-cpp commented on ui/app/mainui/Popups.qml:
```diff
@@ -1253,6 +1253,7 @@ QtObject {
                     swapFormData: swapInputParamsForm
                     swapOutputData: SwapOutputData{}
                 }
+                loginType: root.rootStore.loginType
```
can we please add this to the SwapStore, if there is no correct store for it yet?

caybro replied:
It is coming from the very top (shared) RootStore; do we really need to add it to the SwapStore as well?

caybro replied:
Yeah then I can't see why we should add it here? It's a common property for the whole app

Khushboo-dev-cpp replied:
because all properties that go to Swap and currently going via the store or the adaptor and it would be nice ti keep that consistent.

Khushboo-dev-cpp reviewed (commented):
The popups are looking amazing!!

just a few issues I found visually
1. Dont See keycard icon on the sign button in design maybe its a common behaviour? 
<img width="1061" alt="Screenshot 2024-07-11 at 5 51 54 PM" src="https://github.com/status-im/status-desktop/assets/60327365/715f0350-7f46-4983-acca-4e6f1449cf0d">

<img width="1075" alt="Screenshot 2024-07-11 at 5 57 49 PM" src="https://github.com/status-im/status-desktop/assets/60327365/60e4f317-6a4c-4fbc-bcf1-2fa4981f64cb">

 2. Network icon blurry
 
 3. Kebab menu on on pay side in design https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4187-89431&m=dev
 4. Color mismatch https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4187-89438&m=dev 
 5.  Name mismatch in context menu for network selected
<img width="1075" alt="Screenshot 2024-07-11 at 6 04 23 PM" src="https://github.com/status-im/status-desktop/assets/60327365/d3983cc3-7152-4978-ac5c-4bb31d0adfb8">
6. As per design opus should close after copy https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4069-188707&m=dev
https://github.com/status-im/status-desktop/assets/60327365/9a7b5a44-5c22-4179-8bd2-1bbc88fdc824
7. the link outs from context menu are not working 
https://github.com/status-im/status-desktop/assets/60327365/305a1d91-9ceb-4549-a62c-42b665ece498

caybro commented:
> The popups are looking amazing!!
> 
> just a few issues I found visually
> 
> 1. Dont See keycard icon on the sign button in design maybe its a common behaviour?

Yeah, it should display when the account selected for the swap transaction requires auth using the keycard, see the respective "keycard" icon in the account selector:

![image](https://github.com/status-im/status-desktop/assets/5377645/0046aea7-c5bc-4621-b2d6-67e403031cd7)

> 2. Network icon blurry

Hmm not here... but then maybe I should just use the "tiny" version for the badge :thinking: 

> 3. Kebab menu on on pay side in design https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4187-89431&m=dev

What's wrong with it?

> 4. Color mismatch https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4187-89438&m=dev

You mean the "max. fees" text?

> 5. Name mismatch in context menu for network selected

Ah, it's the test network... will fix

> 6. As per design opus should close after copy 

Opus?

> 8. the link outs from context menu are not working

Will check it out

caybro commented:
> > 3\. Network icon blurry
> 
> Hmm not here... but then maybe I should just use the "tiny" version for the badge 🤔

Ah, definitely not :) This is not how the network icon should look like (compared to Figma):
![image](https://github.com/status-im/status-desktop/assets/5377645/a9356de6-381b-40be-a6bf-90c6865b15ed)

That said, it's an SVG in any case and no reason why it should be blurred (except for running it in a virtual machine ;p

caybro commented:
> > 5\. Kebab menu on on pay side in design https://www.figma.com/design/TS0eQX9dAZXqZtELiwKIoK/Swap---Milestone-1?node-id=4187-89431&m=dev
> 
> What's wrong with it?

Ah, is it that when it's ETH, it doesn't have the subtitle and kebab menu?
![image](https://github.com/status-im/status-desktop/assets/5377645/09bd4a6f-f9f6-4ac0-a85f-46d9e195b867)

caybro commented:
@Khushboo-dev-cpp I hope I addressed all your comments/feedback, pls have a second look :)

caybro commented on ui/app/AppLayouts/Wallet/popups/swap/SwapInputParamsForm.qml:
```diff
@@ -56,9 +56,7 @@ QtObject {
                 !!root.fromTokensKey && !!root.toTokenKey &&
-                (!!root.fromTokenAmount &&
-                 !isNaN(bigIntNumber) &&
-                 SQUtils.AmountsArithmetic.cmp(bigIntNumber, 0) === 1) &&
```
This was `assert`ing on the `cmp` call, because:

```qml
function cmp(amount1, amount2) {
        console.assert(amount1 instanceof Big.Big)
        console.assert(amount2 instanceof Big.Big)
...
```
`0` is not a Big number object

Khushboo-dev-cpp commented:
> @Khushboo-dev-cpp I hope I addressed all your comments/feedback, pls have a second look :)

Hey @caybro can you share a screen shot of how the network icon looks for you on your side? for it really dpesnt look good. 

also the link outs to etherscan  in any of the cases doesn't work.

caybro commented:
> > @Khushboo-dev-cpp I hope I addressed all your comments/feedback, pls have a second look :)
> 
> Hey @caybro can you share a screen shot of how the network icon looks for you on your side? for it really dpesnt look good.

Sure, he it is:
![image](https://github.com/user-attachments/assets/d2bab335-a5ec-48bf-bfc4-9754abca372f)

Khushboo-dev-cpp commented:
> > also the link outs to etherscan in any of the cases doesn't work.
> 
> is the link wrong for you, or does it not open at all in the browser?

doesnt open at all

Khushboo-dev-cpp reviewed (approved):
LGTM!

dlipicar commented on ui/app/AppLayouts/Wallet/popups/swap/SwapApproveCapModal.qml:
```diff
@@ -44,6 +43,7 @@ SignTransactionModalBase {
     property string serviceProviderName: "Paraswap"
     property string serviceProviderURL: Constants.swap.paraswapUrl // TODO https://github.com/status-im/status-desktop/issues/15329
     property string serviceProviderContractAddress: "0x1bD435F3C054b6e901B7b108a0ab7617C808677b"
```
not changed in this PR, but where did this contract address come from?
The one we interact with (approval, swap) is `0x216b4b4ba9f3e719726886d34a177484278bfcae` (fortunately the same address on every chain)
![image](https://github.com/user-attachments/assets/72cc8187-740c-432c-92da-5de955682af4)

dlipicar replied:
Ideally this should be coming from the backend, but in the meantime we probably want to put the write one here :D

dlipicar replied:
Agh, commented too early, it's overriden when instantiating the component. Nevermind!

dlipicar reviewed (approved):
massive work!
